presto-main/src/main/java/com/facebook/presto/operator/exchange/PartitioningExchanger.java

```diff
@@ -62,37 +63,46 @@ public PartitioningExchanger(
         }
         partitionGenerator = new LocalPartitionGenerator(hashGenerator, buffers.size());
 
-        partitionAssignments = new IntList[partitions.size()];
-        for (int i = 0; i < partitionAssignments.length; i++) {
-            partitionAssignments[i] = new IntArrayList();
-        }
+        pageBuilders = IntStream.range(0, buffers.size())
+                .mapToObj(i -> new PageBuilder(types))
+                .collect(toImmutableList());
     }
 
     @Override
     public synchronized void accept(Page page)
     {
-        // reset the assignment lists
-        for (IntList partitionAssignment : partitionAssignments) {
-            partitionAssignment.clear();
-        }
-
-        // assign each row to a partition
         for (int position = 0; position < page.getPositionCount(); position++) {
             int partition = partitionGenerator.getPartition(position, page);
-            partitionAssignments[partition].add(position);
+            PageBuilder pageBuilder = pageBuilders.get(partition);
+            appendRow(pageBuilder, page, position);
         }
 
-        // build a page for each partition
-        Block[] sourceBlocks = page.getBlocks();
-        Block[] outputBlocks = new Block[sourceBlocks.length];
-        for (int partition = 0; partition < buffers.size(); partition++) {
-            List<Integer> positions = partitionAssignments[partition];
-            if (!positions.isEmpty()) {
-                for (int i = 0; i < sourceBlocks.length; i++) {
-                    outputBlocks[i] = sourceBlocks[i].copyPositions(positions);
-                }
+        flush(false);
+    }
+
+    @Override
+    public void flush()
+    {
+        flush(true);
+    }
+
+    private void appendRow(PageBuilder pageBuilder, Page page, int position)
```

Comment:
Are there any benchmarks for exchange performance? Is appending row-by-row any slower than using copyPositions, in exchange for having full pages of output that makes subsequent execution faster?